notes/memory-and-management.md

```diff
@@ -0,0 +1,666 @@
+
+- Strongly Typed: The return of the C++ statement is more type-safe than the C counter part, as it explicitly labels the pointer as a specific type.
+- Automatic Memory Calcuation: You do not need to actually specify how big of a memory size you want.
+- Faster Initialization: The use of the constructor allows for declaration and initialization all in 1 step.
+
+However, the need for any of these stem from the fact that C/C++ lack a complete garbage collector, with C lacking any concept of it at all. A garbage collector is a system that automatically manages a system's memory. In particular, both systems lack a system that resets destroyed variables, which involves the resetting of the corresponding memory to its null, or zero value.
+
+A use of C++ style allocation can avoid this by providing initialization values in the constructor, but this cannot be done in many cases. So, what happens in most cases is that when `malloc()` or `new` is used, the resulting pointer points to memory containing the value of the previous variable that used that memory. 
+
+This presents an issue in many contexts, for example, when you need to make a counter variable. Fortunately, there are many solutions, one of which is to just set the variable to its zero or null value before use. The easiest, however, is to use:
+
+<div style="text-align: center">
+
+ `void* calloc(size_t size);` 
+ 
+ or 
+ 
+ `void* calloc(size_t N, size_t size);`
+
+</div>
+
+This is the same as `malloc()`, except it resets the value of the memory it allocates before returning it. As you can tell, this originates from C and not C++.
+
+Say you also want to change the size of a variable that has previously been allocated. You must then use:
+
+<div style="text-align: center">
+
+`void* realloc(void *ptr, size_t new_size);`
+
+</div>
+
+This is the same as `malloc()` except it will allocate the same memory pointed to by `ptr` and more. It typically gets reassigned to `ptr`.
+
+These are all valid ways to allocate memory to the heap, and note that as Figure 3 points out, the heap grows upward the more memory we put into it, hence the name "heap".
+
+### Deallocating Memory
+
+This MUST always happen after you are done with whatever you allocated. In C, this is achieved with:
+
+<div style="text-align: center">
+
+`void free(void *ptr);`
+
+</div>
+
+In C++, this is achieved using `delete`:
+
+<div style="text-align: center">
+
+`delete ptr;`
+
+</div>
+
+However, if it is an array, then we use `delete[]`:
+
+<div style="text-align: center">
+
+`delete[] ptr;`
+
+</div>
+
+If objects are properly made, you will never need to delete an object since when they fall out of scope, they will be deleted. However, there is more to them concerning memory management than meets the eye. See the section on Classes for more information.
+
+In C++, you may "mix and match" the uses for allocation and deallocation, so you can `malloc` a pointer and then `delete` it, or any other combination you can think of.
```

Comment:
This is kinda sus, the C++ standard technically says that this mixing and matching is "undefined behavior." So in practice I'd always want to `delete` or `delete[]` with `new` and `free` with malloc`

Also probably worth noting that if you do `delete` on an array, you only clear out the first element